streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValues.java

```diff
@@ -46,8 +58,104 @@ public KStreamTransformValuesProcessor(ValueTransformer<V, R> valueTransformer)
         }
 
         @Override
-        public void init(ProcessorContext context) {
-            valueTransformer.init(context);
+        public void init(final ProcessorContext context) {
```

Comment:
Because this is all public API, we will need a KIP for this. For now, I would just do the patch to fix the bug and do a proper redesign as a follow up.

About return on `punctuate`: I am not sure how this would resolve the multi-output issues because you can only return a single value via punctuate. Thus, outputting an arbitrary number of records is not possible as punctuate get's called based on progressing stream-time and if I understand this correctly at most once per processed record. Thus, you could at most emit 2 outputs per input. (All this in independent if you compute the output base on one or multiple input record. I cannot follow the `1-to-m` vs `n-to-m` argument.)

Furthermore, assigning the last seen key seems to be a random key assignment and users will have a hard time to understand this. I would rather just not allow this -- an alternative to throwing an exception might be to just ignore the returned value, but I think throwing is more transparent to the user -- otherwise, it might result in "hidden" bugs of the application ("Why the hack do I loose data?")